openml-lightgbm/lightgbm-provider/standard_code_models/extend_model.py

```diff
@@ -0,0 +1,149 @@
+
+Usage:
+ script <LightGBM_model.txt> <number of trees to add> <output_model_path>
+
+Author: Alberto Ferreira
+Copyright: 2020, Feedzai
+License: Same as the repo.
+"""
+
+import argparse
+import random
+
+class ModelBlocks:
+    def __init__(self, filepath):
+        """
+        Initialize ModelBlocks data structures,
+        parse model file and do self-consistency checks.
+        """
+        self.raw_blocks = ModelBlocks.parse_file_blocks(filepath)
+
+        self.header_block = self.raw_blocks[0]
+        self.tree_blocks = list(filter(ModelBlocks.is_tree_block, self.raw_blocks))
+        self.end_blocks = self.raw_blocks[
+            ModelBlocks.find_end_of_trees_block(self.raw_blocks):
+        ]
+        del self.raw_blocks
+
+        self.N_ORIGINAL_TREES = len(self.tree_blocks)
+        self.check_model_blocks(filepath)
+
+    def check_model_blocks(self, filepath):
+        """
+        Asserts that the file output will be the same as the input before any changes are made.
+        """
+        input_lines  = open(filepath).readlines()
+        output_lines = list(self.yield_model_lines()) # list of output lines
+
+        if len(input_lines) != len(output_lines):
+            raise Exception("Bug detected: Output self-consistency checks failed!")
+
+    @staticmethod
+    def parse_file_blocks(filepath):
+        """
+        Parses the model file to a list of blocks.
+        Each block is a "paragraph" in the model file.
+        """
+        at_block = False
+        blocks = []
+        with open(filepath) as f:
+            block = []
+            for line in f:
+                if line == "\n":
+                    block.append(line)
+                    at_block = False
+                else:
+                    if not at_block: # starting new block
+                        if block:
+                            blocks.append(block)
+                        block = []
+                        at_block = True
+                    block.append(line)
+            if block:
+                blocks.append(block)
+        return blocks
+
+    @staticmethod
+    def is_tree_block(block):
+        """Returns true only for tree blocks."""
+        return block and block[0].startswith("Tree=")
+
+    @staticmethod
+    def tree_size(block):
+        """Returns the size of the tree in the file in bytes."""
+        return sum(map(len, block))
+
+    @staticmethod
+    def find_end_of_trees_block(blocks):
+        """Detect the model block that ends the trees section."""
+        for b, block in enumerate(blocks):
+            if len(block) == 2 and block[0] == "end of trees\n":
+                return b
+        raise Exception("Corrupt file: Coudln't find end of trees block!")
+
+    def add_sampled_tree(self):
+        """
+        Adds a copy of an existing tree to the model as the last tree.
+        Returns the index of the sampled tree.
+        """
+        sample_tree_idx = random.randrange(1, self.N_ORIGINAL_TREES)
```

Comment:
Coday is complaining about this as a low-security issue. You should use os.urandom instead of os.random.
However, I think this isn't a big issue and it affects performance. Therefore I'll approve the review after you resolve the other comments.

I just want to raise up the trade-off between performance vs security here.

Reply:
I actually want the files it generates to be reproducible (pseudo-random). Is there a way to disable this check?

If not I guess I can also cycle integers and sample trees in round-robin fashion -> codacy :(.